packages/ethereum-contracts/contracts/EIP712_Canvas.sol

```diff
@@ -0,0 +1,178 @@
+//SPDX-License-Identifier: Unlicense
+pragma solidity ^0.8.19;
+
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "./CID.sol";
+import "./ECDSA_Verify.sol";
+
+string constant sessionType = "Session(address address,string blockhash,uint256 duration,string publicKey,uint256 timestamp)";
+bytes32 constant sessionTypedDataHash = keccak256(bytes(sessionType));
+
+string constant actionType = "Action(address address,bytes args,string blockhash,string name,uint256 timestamp)";
+bytes32 constant actionTypedDataHash = keccak256(bytes(actionType));
+
+string constant messageSessionType = "Message(uint256 clock,string[] parents,Session payload,string topic)Session(address address,string blockhash,uint256 duration,string publicKey,uint256 timestamp)";
+bytes32 constant messageSessionTypedDataHash = keccak256(bytes(messageSessionType));
+
+string constant messageActionType = "Message(uint256 clock,string[] parents,Action payload,string topic)Action(address address,bytes args,string blockhash,string name,uint256 timestamp)";
+bytes32 constant messageActionTypedDataHash = keccak256(bytes(messageActionType));
+
+bytes32 constant emptyDomainSeparator = keccak256(abi.encode(keccak256("EIP712Domain()")));
+
+contract EIP712_Canvas {
+
+    struct Session {
+        address address_;
+        string blockhash_;
+        uint256 duration;
+        string publicKey;
+        uint256 timestamp;
+    }
+
+    struct Action {
+        address address_;
+        bytes args;
+        string blockhash_;
+        string name;
+        uint256 timestamp;
+    }
```

Comment:
i suppose it's hard to avoid `address_` and `blockhash_` as identifiers?

Reply:
Yeah, `address` is a keyword which can't be used elsewhere. `blockhash` is the name of a built in function but we can actually use it as a variable, so I've replaced that.